### PR TITLE
Adjust RNG seeding to pass CAVP testing

### PR DESCRIPTION
**Pull request by paulidale**
NIST CAVP tests require all of the entropy provided to be fed into each DRBG.  Our implementation only fed in what the DRBG thought it required.  In order to pass the tests, we now consume everything that was given.

Outside of testing, this change makes no difference to the seed material that DRBGs receive.  However, it does restore the possibility of us supporting entropy sources that produce less than eight bits of randomness per byte.

- [x] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

File: doc/man7/provider-rand.pod [hunk 1]
```diff
@@ -110,6 +116,19 @@ OSSL_FUNC_rand_nonce() is used to generate a nonce of the given I<strength> with
 length from I<min_noncelen> to I<max_noncelen>. If the output buffer I<out> is
 NULL, the length of the nonce should be returned.
 
+OSSL_FUNC_rand_get_seed() is used by deterministic generators to
+obtain their seeding material from their parent.  The seed bytes will meet
+the specified
+security level of I<entropy> bits and there will be between I<min_len>
```
Comment (slontis) [inline doc/man7/provider-rand.pod:122]: strange line spacing here.

Comment (paulidale) [reply doc/man7/provider-rand.pod:121]: It's a POD file, it doesn't matter :)

I've reformatted the paragraph.

**Comment by paulidale** [comment]
ping for review.

File: providers/implementations/rands/drbg.c [hunk 2]
```diff
@@ -212,8 +188,8 @@ static size_t prov_drbg_get_entropy(PROV_DRBG *drbg, unsigned char **pout,
  * Implements the cleanup_entropy() callback
```
Comment (slontis) [inline providers/implementations/rands/drbg.c:188]: single line comment now?

Comment (paulidale) [reply providers/implementations/rands/drbg.c]: Fixed.

File: providers/implementations/rands/seed_src.c [hunk 3]
```diff
@@ -170,6 +172,42 @@ static int seed_src_verify_zeroization(ossl_unused void *vseed)
...
+static size_t seed_get_seed(void *vseed, unsigned char **pout,
+                            int entropy, size_t min_len, size_t max_len,
+                            int prediction_resistance,
+                            const unsigned char *adin, size_t adin_len)
+{
+    size_t bytes_needed;
+    unsigned char *p;
+
+    /* Figure out how many bytes we need */
+    bytes_needed = entropy >= 0 ? (entropy + 7) / 8 : 0;
```
Comment (slontis) [inline providers/implementations/rands/seed_src.c:184]: Is this assuming 8 bits of strength per byte??

File: providers/implementations/rands/drbg.c [hunk 4]
```diff
@@ -231,8 +207,42 @@ static size_t get_entropy(PROV_DRBG *drbg, unsigned char **pout, int entropy,
...
+    /*
+     * Get random data from parent.  Include our DRBG address as
+     * additional input, in order to provide a distinction between
+     * different DRBG child instances.
+     *
+     * Note: using the sizeof() operator on a pointer triggers
+     *       a warning in some static code analyzers, but it's
+     *       intentional and correct here.
+     */
+    bytes = drbg->parent_get_seed(drbg->parent, pout, drbg->strength,
```
Comment (slontis) [inline providers/implementations/rands/drbg.c:241]: I assume there was a reason why all the methods of the parent are resolved in the child?
Why doesnt it do drbg->parent->get_seed()

File: providers/implementations/rands/drbg.c [hunk 5]
```diff
@@ -180,26 +165,17 @@ static size_t prov_drbg_get_entropy(PROV_DRBG *drbg, unsigned char **pout,
...
      *       a warning in some static code analyzers, but it's
      *       intentional and correct here.
      */
-    r = drbg->parent_generate(drbg->parent, buffer, bytes_needed,
-                              drbg->strength, prediction_resistance,
-                              (unsigned char *)&drbg,
-                              sizeof(drbg));
-    ossl_drbg_unlock_parent(drbg);
-    if (r == 0) {
+    if (!ossl_prov_drbg_generate(drbg, buffer, bytes_needed,
```
Comment (slontis) [inline providers/implementations/rands/drbg.c:176]: where is parent_generate() used now??

File: providers/implementations/rands/seed_src.c (hunk 3 above)
Comment (paulidale) [reply providers/implementations/rands/seed_src.c:184]: Yes.  For other seed sources, it need not be true but for ours it's a safe assumption.

File: providers/implementations/rands/drbg.c (hunk 4 above)
Comment (paulidale) [reply providers/implementations/rands/drbg.c:241]: There are a couple of reasons:

1. The DRBG's parent is an opaque blob of data that could live in a different provider.
2. More subtle is that parent's _context_ doesn't contain pointers to the parent's functions.  The only way they are found is via the `OSSL_DISPATCH` array that the fetch hands back.  In EVP terms, what's being accessed would be drbg->parent->meth->get_seed().  On the provider side, there is no equivalent to the _meth_ pointer in the parent (plus the parent is an opaque blob).

File: providers/implementations/rands/drbg.c (hunk 5 above)
Comment (paulidale) [reply providers/implementations/rands/drbg.c:176]: I don't think it is anymore.  Have removed.

**Comment by slontis** [comment]
A few more things probably need to be removed??
OSSL_FUNC_RAND_GENERATE and the docs for this?

**Comment by paulidale** [comment]
The generate call is still used, just not for the child to grab seed material from its parent.

File: providers/implementations/rands/seed_src.c (hunk 3 above)
Comment (slontis) [reply providers/implementations/rands/seed_src.c:184]: Did you push the comment for this?

Comment (paulidale) [reply providers/implementations/rands/seed_src.c:189]: I have now.

**Review by slontis — APPROVED** [review]

**Comment by paulidale** [comment]
Merged to master, thanks for the review comments.

**Pull request closed without merging**